File: web/lib/agentRunner.ts
import { spawn, execSync } from "child_process";
import path from "path";
import fs from "fs";

function findClaude(): string {
  const isWindows = process.platform === "win32";
  const localBin = path.join(process.cwd(), "node_modules", ".bin", isWindows ? "claude.exe" : "claude");
  const pkgBin = path.join(
    process.cwd(), "node_modules", "@anthropic-ai", "claude-code", "bin",
    isWindows ? "claude.exe" : "claude",
  );
  const platformPkg = path.join(
    process.cwd(), "node_modules", "@anthropic-ai",
    `claude-code-${process.platform}-${process.arch}`,
    isWindows ? "claude.exe" : "claude",
  );
  for (const p of [pkgBin, platformPkg, localBin]) {
    if (fs.existsSync(p)) return p;
  }
  try {
    const lines = execSync(isWindows ? "where claude" : "which claude", { encoding: "utf8" })
      .trim().split("\n");
    const exe = isWindows ? lines.find((l) => l.trim().toLowerCase().endsWith(".exe")) : undefined;
    return (exe ?? lines[0]).trim();
  } catch {
    return isWindows ? "claude.exe" : "claude";
  }
}

export const CLAUDE_BIN = findClaude();
console.log("[findClaude]", CLAUDE_BIN, "| exists:", fs.existsSync(CLAUDE_BIN));

export const WIKI_DIR = path.resolve(process.cwd(), "../../wiki-llm");

// CEO 응답 대기 — 세션별 Promise resolver 저장
export const pendingResponses = new Map<string, (r: string) => void>();

// 취소된 세션 ID 집합
export const cancelledSessions = new Set<string>();

export type RunAgentOptions = {
  allowedTools?: string[];
  noTools?: boolean;
  addDirs?: string[];
  cwd?: string;
  maxTurns?: number;
};

function formatToolUse(name: string, input: Record<string, unknown>): string {
  switch (name) {
    case "WebSearch": return `\n🔍 검색: "${input.query ?? ""}"\n`;
    case "WebFetch":  return `\n🌐 페이지 읽는 중: ${input.url ?? ""}\n`;
    case "Read":      return `\n📖 파일 읽는 중: ${input.file_path ?? ""}\n`;
    case "Write":     return `\n✏️ 파일 작성: ${input.file_path ?? ""}\n`;
    case "Edit":      return `\n✂️ 파일 수정: ${input.file_path ?? ""}\n`;
    case "Grep":      return `\n🔎 검색: "${input.pattern ?? ""}"\n`;
    case "Glob":      return `\n📂 파일 탐색: ${input.pattern ?? ""}\n`;
    default:          return `\n⚙️ ${name}\n`;
  }
}

export async function runAgent(
  prompt: string,
  options: RunAgentOptions = {},
  onProgress?: (text: string) => void,
  onThinking?: (text: string) => void,
): Promise<string> {
  return new Promise((resolve, reject) => {
    const args = [
      "-p",
      "--verbose",
      "--output-format", "stream-json",
      "--include-partial-messages",
      "--dangerously-skip-permissions",
    ];
    if (options.maxTurns != null) {
      args.push("--max-turns", String(options.maxTurns));
    }
    if (options.noTools) {
      args.push("--tools", "");
    } else if (options.allowedTools?.length) {
      args.push("--allowedTools", options.allowedTools.join(","));
    }
    for (const dir of options.addDirs ?? []) {
      if (fs.existsSync(dir)) args.push("--add-dir", dir);
    }

    const proc = spawn(CLAUDE_BIN, args, {
      cwd: options.cwd ?? process.cwd(),
      env: process.env,
    });

    proc.stdin.write(prompt);
    proc.stdin.end();

    let buffer = "";
    let finalResult = "";
    let lastText = "";

    proc.stdout.on("data", (chunk: Buffer) => {
      buffer += chunk.toString();
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        if (!line.trim()) continue;
        try {
          const event = JSON.parse(line.trim());
          if (event.type === "result" && event.subtype === "success") {
            finalResult = event.result ?? "";
          } else if (event.type === "stream_event") {
            const se = event.event;
            if (se?.type === "content_block_delta") {
              const delta = se.delta;
              if (delta?.type === "text_delta" && delta.text && onProgress) {
                onProgress(delta.text as string);
              } else if (delta?.type === "thinking_delta" && delta.thinking && onThinking) {
                onThinking(delta.thinking as string);
              }
            }
          } else if (event.type === "assistant" && event.message?.content) {
            for (const block of event.message.content) {
              if (block.type === "text" && block.text) {
                lastText = block.text as string;
              } else if (block.type === "tool_use" && onProgress) {
                const toolLine = formatToolUse(block.name as string, block.input as Record<string, unknown>);
                if (toolLine) onProgress(toolLine);
              }
            }
          }
        } catch { /* ignore malformed lines */ }
      }
    });

    proc.stderr.on("data", (chunk: Buffer) => {
      console.error("[claude stderr]", chunk.toString());
    });

    proc.on("error", (err) => {
      console.error("[claude spawn error]", err.message, "| bin:", CLAUDE_BIN);
      reject(err);
    });

    proc.on("close", (code) => {
      const result = finalResult || lastText;
      if (code === 0 || result) resolve(result);
      else reject(new Error(`Agent process exited with code ${code}`));
    });
    proc.on("error", reject);
  });
}

export function parseJSON<T>(text: string, fallback: T): T {
  const match = text.match(/```(?:json)?\s*([\s\S]*?)```/) ?? text.match(/(\{[\s\S]*\})/);
  try {
    return JSON.parse(match?.[1] ?? text) as T;
  } catch {
    return fallback;
  }
}

File: backend/orchestrator/agent_runner.py
import asyncio
import json
import os
import shutil
import sys
from pathlib import Path
from typing import Callable


def find_claude() -> str:
    backend_dir = Path(__file__).parent.parent
    web_dir = backend_dir.parent / "web"
    nm = web_dir / "node_modules"

    if sys.platform == "win32":
        candidates = [
            nm / "@anthropic-ai" / "claude-code-win32-x64" / "claude.exe",
            nm / "@anthropic-ai" / "claude-code" / "bin" / "claude.exe",
            nm / ".bin" / "claude.exe",
        ]
        fallback = shutil.which("claude.exe") or shutil.which("claude") or "claude.exe"
    else:
        candidates = [
            nm / ".bin" / "claude",
            nm / "@anthropic-ai" / "claude-code" / "bin" / "claude",
            nm / "@anthropic-ai" / "claude-code-darwin-arm64" / "bin" / "claude",
            nm / "@anthropic-ai" / "claude-code-darwin-x64" / "bin" / "claude",
            nm / "@anthropic-ai" / "claude-code-linux-x64" / "bin" / "claude",
        ]
        fallback = shutil.which("claude") or "claude"

    for p in candidates:
        if p.exists():
            return str(p)
    return fallback


CLAUDE_BIN = find_claude()
WIKI_DIR = str(Path(__file__).parent.parent.parent / "wiki-llm")

print(f"[agent_runner] claude binary: {CLAUDE_BIN}")


def format_tool_use(name: str, input_data: dict) -> str:
    mapping = {
        "WebSearch": lambda i: f"\n🔍 검색: \"{i.get('query', '')}\"\n",
        "WebFetch": lambda i: f"\n🌐 페이지 읽는 중: {i.get('url', '')}\n",
        "Read": lambda i: f"\n📖 파일 읽는 중: {i.get('file_path', '')}\n",
        "Write": lambda i: f"\n✏️ 파일 작성: {i.get('file_path', '')}\n",
        "Grep": lambda i: f"\n🔎 검색: \"{i.get('pattern', '')}\"\n",
        "Glob": lambda i: f"\n📂 파일 탐색: {i.get('pattern', '')}\n",
    }
    fn = mapping.get(name)
    return fn(input_data) if fn else f"\n⚙️ {name}\n"


async def _drain_stderr(proc) -> None:
    """stderr 버퍼를 비워 subprocess 블로킹 방지."""
    if proc.stderr:
        try:
            await proc.stderr.read()
        except Exception:
            pass


async def run_agent(
    prompt: str,
    allowed_tools: list[str] | None = None,
    no_tools: bool = False,
    add_dirs: list[str] | None = None,
    cwd: str | None = None,
    max_turns: int | None = None,
    on_stream: Callable[[str], None] | None = None,
    should_stop: Callable[[], bool] | None = None,
) -> tuple[str, str]:
    """Claude CLI를 asyncio 서브프로세스로 실행. stdout을 실시간 라인 단위로 읽어 on_stream 호출.
    should_stop()이 True를 반환하면 subprocess를 즉시 kill하고 지금까지 받은 결과를 반환."""
    args = [
        CLAUDE_BIN, "-p",
        "--verbose",
        "--output-format", "stream-json",
        "--include-partial-messages",
        "--dangerously-skip-permissions",
    ]
    if max_turns is not None:
        args += ["--max-turns", str(max_turns)]
    if no_tools:
        args += ["--tools", ""]
    elif allowed_tools:
        args += ["--allowedTools", ",".join(allowed_tools)]
    for d in (add_dirs or []):
        if os.path.exists(d):
            args += ["--add-dir", d]

    work_dir = cwd or str(Path(__file__).parent.parent)

    proc = await asyncio.create_subprocess_exec(
        *args,
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        cwd=work_dir,
    )

    assert proc.stdin is not None
    proc.stdin.write(prompt.encode("utf-8"))
    await proc.stdin.drain()
    proc.stdin.close()

    final_result = ""
    last_text = ""
    stream_chunks: list[str] = []

    assert proc.stdout is not None
    try:
        async with asyncio.timeout(600):
            async for raw_line in proc.stdout:
                line = raw_line.decode("utf-8", errors="replace").strip()
                if not line:
                    continue
                try:
                    event = json.loads(line)
                    etype = event.get("type")

                    if etype == "result":
                        r = event.get("result", "") or ""
                        if r:
                            final_result = r

                    elif etype == "stream_event":
                        se = event.get("event", {})
                        if se.get("type") == "content_block_delta":
                            delta = se.get("delta", {})
                            if delta.get("type") == "text_delta":
                                chunk = delta.get("text", "")
                                if chunk:
                                    stream_chunks.append(chunk)
                                    if on_stream:
                                        on_stream(chunk)

                    elif etype == "assistant":
                        # 한 메시지 내 텍스트 블록 전부 합산 (마지막 블록만 유지하면 중간 JSON 유실)
                        texts = [
                            b.get("text", "")
                            for b in event.get("message", {}).get("content", [])
                            if b.get("type") == "text"
                        ]
                        combined = "\n".join(t for t in texts if t)
                        if combined:
                            last_text = combined
                        for block in event.get("message", {}).get("content", []):
                            if block.get("type") == "tool_use":
                                tool_line = format_tool_use(block.get("name", ""), block.get("input", {}))
                                if tool_line:
                                    stream_chunks.append(tool_line)
                                    if on_stream:
                                        on_stream(tool_line)
                except Exception:
                    pass

                if should_stop and should_stop():
                    proc.kill()
                    break
    except TimeoutError:
        proc.kill()

    await asyncio.gather(proc.wait(), _drain_stderr(proc))
    # stream_chunks는 tool-use 에이전트의 신뢰할 수 있는 fallback
    return final_result or last_text or "".join(stream_chunks), "".join(stream_chunks)


def parse_json(text: str, fallback):
    import re
    # 마지막 ```json 코드블록 우선
    blocks = list(re.finditer(r"```(?:json)?\s*([\s\S]*?)```", text))
    if blocks:
        try:
            return json.loads(blocks[-1].group(1).strip())
        except Exception:
            pass
    # 마지막 { 위치부터 시도
    last_brace = text.rfind("{")
    if last_brace != -1:
        try:
            return json.loads(text[last_brace:])
        except Exception:
            pass
    # greedy {…} 추출
    match = re.search(r"(\{[\s\S]*\})", text)
    try:
        return json.loads(match.group(1) if match else text)
    except Exception:
        return fallback
